docs/AllocatorClaims.md

```diff
@@ -0,0 +1,120 @@
+
+Additional constraints
+----------------------
+
+In addition to solving the basic problem, we have a number of additional constraints on the design space:
+
+ - A solution should be possible to apply at the boundaries when wrapping an existing component in a compartment.
+ - Most objects will not use this mechanism (they will be reachable from a single compartment) and so there should be no overhead when not in use.
+ - We are targeting resource-constrained systems and so the total overhead must be low.
+ - We assume compartments may be malicious and so they should not be able to trick the allocator into creating a data structure that takes a very large amount of time to walk.
+
+Possible approach 0: Explicit copies
+------------------------------------
+
+The first approach is simply to punt on the problem entirely.
+When an OS kernel interacts with userspace memory, it calls explicit helpers to copy data to and from the kernel.
+This is necessary because the userspace process may provide invalid pointers or update the memory map to invalidate pointers while the kernel runs.
+
+In many situations, the same approach could work with CHERIoT.
+We could provide a safe memcpy as a library function that runs with interrupts disabled, checks the source and destination pointers, and then performs the copy (up to some bounded size) and reports whether the copy succeeded.
+This would be sufficient in a lot of cases but is not a very friendly programmer model.
+In particular, it means that any compartment wrapping existing unmodified code would have to either (deep) copy any objects passed in or would have to make invasive changes to the wrapped code.
+
+Note: This approach can be orthogonal to others and so we should implement it anyway.
+
+Possible approach 1: Hazard pointers
+------------------------------------
+
+Hazard pointers provide inspiration for the first possible solution.
+This approach involves having a per-compartment list of held objects that the allocator must consult before freeing an object.
+The allocator would provide an API that allowed a compartment to register a table containing a (bounded) array of held objects and an allocation capability.
+On deallocation, the allocator would need to traverse all such lists to determine whether an object is on any of them.
+If the object is found on a list then the metadata should be updated to mark it as owned by the allocation capability associated with the list.
+
+This is attractive for the fast-path case, because the compartment wishing to claim an object for the duration of a call needs only to insert a pointer to it into a pre-registered array.
+It is unclear whether it is possible to use this interface securely.
+In particular, a malicious compartment could allocate a large object and pass a pointer to a small sub-object to the callee.
+The callee has no mechanism (other than calling the allocator) to know whether it holds a pointer to an entire object.
```

Comment:
The allocators traversal could do a subobject check, although we could really do with an `issubset` instruction for that.

Reply:
The CHERIoT `malloc` has no particularly convenient way to map from a sub-object to its allocation, since it's a `dlmalloc`-style inline structure.  It'd be O(n) to walk the heap, uncomfortable with or tricky without holding the lock. :(

Reply:
Oh, right, yes, I forgot that we marked headers that way.